src/runner_metrics.py

```diff
@@ -0,0 +1,222 @@
+        raise CorruptMetricDataError(
+            f"Pre-job metrics for runner {fs.runner_name} is not a JSON object."
+        )
+
+    if not isinstance(post_job_metrics, dict) and post_job_metrics is not None:
+        raise CorruptMetricDataError(
+            f"Post-job metrics for runner {fs.runner_name} is not a JSON object."
+        )
+
+    try:
+        return RunnerMetrics(
+            installed_timestamp=installed_timestamp,
+            pre_job=PreJobMetrics(**pre_job_metrics),
+            post_job=PostJobMetrics(**post_job_metrics) if post_job_metrics else None,
+        )
+    except ValidationError as exc:
+        raise CorruptMetricDataError(str(exc)) from exc
+
+
+def _issue_runner_metrics(runner_metrics: RunnerMetrics, flavor: str) -> None:
+    """Issue metrics.
+
+    Converts the metrics into respective metric events and issues them.
+
+    Args:
+        runner_metrics: The metrics to be issued.
+        flavor: The flavor of the runners.
+    """
+    event = metrics.RunnerStart(
+        timestamp=runner_metrics.pre_job.timestamp,
+        flavor=flavor,
+        workflow=runner_metrics.pre_job.workflow,
+        repo=runner_metrics.pre_job.repository,
+        github_event=runner_metrics.pre_job.event,
+        idle=runner_metrics.pre_job.timestamp - runner_metrics.installed_timestamp,
+    )
+    metrics.issue_event(event)
+
+
+def _clean_up_shared_fs(fs: shared_fs.SharedFilesystem) -> None:
+    """Clean up the shared filesystem.
+
+    Remove all metric files and afterwards the shared filesystem.
+    Args:
+        fs: The shared filesystem for a specific runner.
+    """
+    fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME).unlink(missing_ok=True)
+    fs.path.joinpath(POST_JOB_METRICS_FILE_NAME).unlink(missing_ok=True)
+    fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME).unlink(missing_ok=True)
+
+    try:
+        shared_fs.delete(fs.runner_name)
+    except errors.DeleteSharedFilesystemError:
+        logger.exception("Could not delete shared filesystem for runner %s.", fs.runner_name)
+
+
+def extract(flavor: str, ignore_runners: set[str]) -> None:
+    """Extract and issue metrics from runners.
+
+    The metrics are extracted from the shared filesystem of given runners
+    and respective metric events are issued.
+    Orphan shared filesystems are cleaned up.
+
+    If corrupt data is found, an error is raised immediately, as this may indicate that a malicious
+    runner is trying to manipulate the shared file system.
+    In order to avoid DoS attacks, the file size is also checked.
+
+    Args:
+        flavor: The flavor of the runners to extract metrics from.
+        ignore_runners: The set of runners to ignore.
+
+    Raises:
+        CorruptMetricDataError: If one of the files inside the shared filesystem is not valid.
+    """
+    for fs in shared_fs.list_all():
+        if fs.runner_name not in ignore_runners:
+            metrics_from_fs = _extract_metrics_from_fs(fs)
```

Comment:
Shouldn't we catch exceptions here, because if any runner fail, all the other won't be treated

Reply:
Hm, the only "expected" exception would be `CorruptMetricDataError`. I am not sure if a catch would be too quiet as this could be a security issue (malicious runner) or a programming error.